File: julia/21_one-qubit.jl
#!/usr/bin/env julia
# One-qubit gate examples

function main()
    # |0>, |1>
    ket0 = reshape(ComplexF64[1, 0], 2, 1)
    ket1 = reshape(ComplexF64[0, 1], 2, 1)
    println("|0> =\n$(ket0)\n")
    println("|1> =\n$(ket1)\n")

    # X gate
    X = ComplexF64[0 1; 1 0]
    println("X =\n$(X)\n")
    println("X|0> = $(X * ket0)\n")
    println("X|1> = $(X * ket1)\n")

    # Y gate
    Y = ComplexF64[0 -1im; 1im 0]
    println("Y =\n$(Y)\n")
    println("Y|0> = $(Y * ket0)\n")
    println("Y|1> = $(Y * ket1)\n")

    # Z gate
    Z = ComplexF64[1 0; 0 -1]
    println("Z =\n$(Z)\n")
    println("Z|0> = $(Z * ket0)\n")
    println("Z|1> = $(Z * ket1)\n")

    # Hadamard gate
    H = ComplexF64[1 1; 1 -1] / sqrt(2)
    println("H =\n$(H)\n")
    println("H|0> = $(H * ket0)\n")
    println("H|1> = $(H * ket1)\n")
end

main()

File: python/21_one-qubit.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
One-qubit gate examples
"""

import numpy as np


def main():
    # |0>, |1>
    ket0 = np.array([[1], [0]], dtype=complex)
    ket1 = np.array([[0], [1]], dtype=complex)
    print(f"|0> =\n{ket0}\n")
    print(f"|1> =\n{ket1}\n")

    """X gate"""
    X = np.array([[0, 1], [1, 0]], dtype=complex)
    print(f"X =\n{X}\n")
    print(f"X|0> = {X @ ket0}\n")
    print(f"X|1> = {X @ ket1}\n")

    """Y gate"""
    Y = np.array([[0, -1j], [1j, 0]], dtype=complex)
    print(f"Y =\n{Y}\n")
    print(f"Y|0> = {Y @ ket0}\n")
    print(f"Y|1> = {Y @ ket1}\n")

    """Z gate"""
    Z = np.array([[1, 0], [0, -1]], dtype=complex)
    print(f"Z =\n{Z}\n")
    print(f"Z|0> = {Z @ ket0}\n")
    print(f"Z|1> = {Z @ ket1}\n")

    """Hadamard gate"""
    H = np.array([[1, 1], [1, -1]], dtype=complex) / np.sqrt(2)
    print(f"H =\n{H}\n")
    print(f"H|0> = {H @ ket0}\n")
    print(f"H|1> = {H @ ket1}\n")


if __name__ == "__main__":
    main()
